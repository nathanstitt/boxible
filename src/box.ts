import styled, { css } from 'styled-components'
import {
    edgeStyle, overflowStyle, genericStyles, GenericPropsI, Area, SizeT,
    SIZES, ALIGN_MAP, ALIGN_CONTENT, BASIS, JUSTIFY, FLEX,
} from './styles'

const basisStyle = (basis: string | keyof typeof BASIS) => (
    `flex-basis: ${BASIS[basis] || basis}`
)

type directionT = 'column' | 'row'

const directionStyle = (direction: directionT) => {
    return css`
        min-width: 0;
        min-height: 0;
        flex-direction: ${direction};
    `
}


interface FlexGrowShrinkI {
    grow?: string
    shrink?: string
}

type FlexGrowT = boolean | keyof typeof FLEX | FlexGrowShrinkI

const flexStyle = (flex: FlexGrowT, basis?: string | keyof typeof BASIS) => {
    let flexStyle = ''
    if (typeof flex === 'boolean') {
        flexStyle = flex ? '1 1' : '0 0'
    }

    if (typeof flex == 'string') {
        flexStyle = FLEX[flex]
    }

    const basisStyle = flex !== true && !basis ? ' auto' : ''

    return `flex: ${flexStyle}${basisStyle}; `
}

const fillStyle = (fillProp: string | boolean) => {
    if (fillProp === 'horizontal') {
        return 'width: 100%;'
    }
    if (fillProp === 'vertical') {
        return 'height: 100%;'
    }
    if (fillProp) {
        return css`
            width: 100 %;
            height: 100 %;
        `
    }
    return undefined
}

const justifyStyle = (justify: keyof typeof JUSTIFY) => (
    css`justify-content: ${JUSTIFY[justify]};`
)

const WRAP_MAP = {
    reverse: 'wrap-reverse',
}

type WrapT = boolean | keyof typeof WRAP_MAP

const wrapStyle = (wrap: WrapT) => {
    if (typeof wrap === 'boolean') {
        return wrap ? css`flex-wrap: wrap` : undefined
    }
    return css`flex-wrap: ${WRAP_MAP[wrap]}`
}

interface MinMaxI {
    max?: string
    min?: string
}

const widthStyle = (w: string | MinMaxI) => {
    if (typeof w === 'object') {
        const c: any = {}
        if (w.max) c.maxWidth = w.max
        if (w.min) c.minWidth = w.min
        return css(c)
    } else {
        return css`width: ${w}`
    }
}

const heightStyle = (w: string | MinMaxI) => {
    if (typeof w === 'object') {
        const c: any = {}
        if (w.max) c.maxHeight = w.max
        if (w.min) c.minHeight = w.min
        return css(c)
    } else {
        return css`height: ${w}`
    }
}

const gapStyle = (gapProp: string | true) => {
    let gap = gapProp
    if (typeof gapProp === 'boolean') {
        if (gap) {
            gap = 'default'
        } else {
            return
        }
    }
    const size = SIZES[gap as string] || gap
    return css`
        > *:not(:last-child) { margin-right: ${size}; }
    `
}

export interface BoxProps extends GenericPropsI {
    align?: keyof typeof ALIGN_MAP
    alignContent?: keyof typeof ALIGN_CONTENT
    direction?: directionT
    flex?: FlexGrowT
    basis?: string | keyof typeof BASIS
    justify?: keyof typeof JUSTIFY
    gap?: boolean | SizeT
    height?: string | MinMaxI
    width?: string | MinMaxI
    fill?: boolean | 'horizontal' | 'vertical'
    wrap?: WrapT
    className?: string
    id?: string | number
    pad?: SizeT | Area
}

const DOM_PROPS = ['children', 'onClick', 'type', 'role', 'id', 'tabIndex']

// NOTE: basis must be after flex! Otherwise, flex overrides basis
export const Box = styled.div
    .withConfig({
        shouldForwardProp: (prop) => prop.startsWith('data') || DOM_PROPS.includes(prop),
    }) <BoxProps>`
    display: flex;
    box-sizing: border-box;
    outline: none;
    ${props => !props.basis && 'max-width: 100%;'};
    ${props => props.align && `align-items: ${ALIGN_MAP[props.align]};`}
    ${props => props.alignContent && `align-content: ${ALIGN_CONTENT[props.alignContent]};`}
    ${props => props.direction && directionStyle(props.direction)}
    ${(props: any) => props.overflowProp && overflowStyle(props.overflowProp)}
    ${({ flex, basis }) => flex && flexStyle(flex, basis)}
    ${({ basis }) => basis && basisStyle(basis)}
    ${({ justify }) => justify && justifyStyle(justify)}
    ${({ gap }: any) => gap && gapStyle(gap)}
    ${({ height }) => height && heightStyle(height)}
    ${({ width }) => width && widthStyle(width)}
    ${({ fill }) => fill && fillStyle(fill)}
    ${({ wrap }) => wrap && wrapStyle(wrap)}
    ${props => props.pad && css(edgeStyle('padding', props.pad))}

    ${props => css(genericStyles(props))}
`
