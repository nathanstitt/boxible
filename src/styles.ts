import { CSSObject } from '@emotion/react'
import { Theme } from '@emotion/react'
import { capitalize } from './util'
import type { Property } from 'csstype'

export const SIZES = {
    small: '.2rem',
    default: '.5rem',
    medium: '.8rem',
    large: '1rem',
    xlarge: '2.5rem',
    xxlarge: '4rem',
}

export type edgeKindT = "margin" | "border" | "padding"
export type Size = keyof typeof SIZES

export const SCREEN_SIZES = {
    xs: `max-width: 576px`,
    sm: `min-width: 576px`,
    md: `min-width: 768px`,
    lg: `min-width: 992px`,
    xl: `min-width: 1200px`,
    xxl: `min-width: 1400px`,
    mobile: `max-width: 767px`,
    tablet: `min-width: 768px`,
    desktop: `min-width: 992px`,
}
export type ScreenSizeNames = keyof typeof SCREEN_SIZES

export const ALIGN_CONTENT_MAP = {
    start: 'flex-start',
    center: 'center',
    end: 'flex-end',
    around: 'space-around',
    between: 'space-between',
    evenly: 'space-evenly',
    stretch: 'stretch',
}
//export type AlignContentNames = keyof typeof ALIGN_CONTENT

export const BASIS = {
    '1/2': '50%',
    '1/4': '25%',
    '2/4': '50%',
    '3/4': '75%',
    auto: 'auto',
    full: '100%',
    '1/3': '33.33%',
    '2/3': '66.66%',
}

export const JUSTIFY_MAP = {
    start: 'flex-start',
    center: 'center',
    end: 'flex-end',
    around: 'space-around',
    between: 'space-between',
    evenly: 'space-evenly',
}

export const FLEX = {
    grow: '1 0',
    shrink: '0 1',
}

export interface Area {
    horizontal?: Size | string | number
    vertical?: Size | string | number
    top?: Size | string | number
    bottom?: Size | string | number
    left?: Size | string | number
    right?: Size | string | number
}

export type Side = keyof Area

export const ALIGN_SELF_MAP = {
    center: 'center',
    end: 'flex-end',
    start: 'flex-start',
    stretch: 'stretch',
    left: 'left',
    right: 'right',
    baseline: 'baseline',
}


export const ALIGN_MAP = {
    baseline: 'baseline',
    center: 'center',
    end: 'flex-end',
    start: 'flex-start',
    stretch: 'stretch',
}

export type AlignNames = keyof typeof ALIGN_MAP

export interface GenericProps {
    margin?: number | string | Size | Area | Side
    padding?: number | string | Size | Area | Side
}

export const genericStyles = (props: GenericProps & { theme: Theme }): CSSObject => {
    const styles: CSSObject = {}
    if (props.margin) {
        Object.assign(styles, edgeStyle('margin', props.margin))
    }
    if (props.padding) {
        Object.assign(styles, edgeStyle('padding', props.padding))
    }
    return styles
}


interface OverflowI {
    horizontal?: Property.OverflowX
    vertical?: Property.OverflowX
}

export const overflowStyle = (overflowProp: string | OverflowI): CSSObject => {
    if (typeof overflowProp === 'string') {
        return { overflow: overflowProp }
    }
    const style: CSSObject = {}
    if (overflowProp.horizontal) style['overflowX'] = overflowProp.horizontal;
    if (overflowProp.vertical) style['overflowY'] = overflowProp.vertical;
    return style
}

const SIDES = ['top', 'right', 'left', 'bottom', 'horizontal', 'vertical']

export const edgeStyle = (
    kind: edgeKindT,
    data: string | number | Size | Area,
): CSSObject => {
    if (typeof data === 'string') {
        if (SIDES.includes(data)) {
            data = { [`${data}`]: SIZES.default }
        } else {
            return { [kind]: SIZES[data] || data }
        }
    }
    if (typeof data == 'number') {
        return { [kind]: `${data}px` }
    }

    const styles: CSSObject = {}

    if (data.horizontal) {
        styles[`${kind}Right`] = styles[`${kind}Left`] = SIZES[data.horizontal] || data.horizontal
    }
    if (data.vertical) {
        styles[`${kind}Top`] = styles[`${kind}Bottom`] = SIZES[data.vertical] || data.vertical
    }

    ['top', 'right', 'bottom', 'left'].forEach(side => {
        const value = data[side]
        if (value) {
            styles[`${kind}${capitalize(side)}`] = SIZES[value] || value
        }
    })

    return styles
}
