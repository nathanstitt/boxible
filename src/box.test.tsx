/* global describe, it, expect */
import React from 'react'
import renderer from 'react-test-renderer'
import { matchers } from '@emotion/jest'
import { Box } from './box'
import { SIZES, SCREEN_SIZES } from './styles'


// Add the custom matchers provided by '@emotion/jest'
expect.extend(matchers)

describe('Box Component', () => {
    beforeEach(() => {
        SIZES.large = '1rem'
        SCREEN_SIZES.lg = 'min-width: 992px'
    })

    it('renders and matches snapshot', () => {
        const tree = renderer.create(
            <Box
                align="stretch"
                alignContent="end"
                direction="columnReverse"
                flex={{ grow: 2, shrink: 1 }}
                basis="1/2"
                justify="evenly"
                gap="medium"
                height={{ min: '20%' }}
                width={{ max: '80%' }}
                fill="vertical"
                wrap
                className="a-box-for-testing"
                id="1"
                pad="large"
                margin="small"
            >
                A test
            </Box>
        ).toJSON();
        expect(tree).toMatchSnapshot()
        expect(tree).toHaveStyleRule('margin', '.2rem')
        expect(tree).toHaveStyleRule('padding', '1rem')
        expect(tree).toHaveStyleRule('flex-wrap', 'wrap')
    })

    it('can set shrink/grow/basis at once', () => {
        const tree = renderer.create(
            <Box flex={{ grow: 3, shrink: false, basis: '1/2' }}>something</Box>
        ).toJSON()
        expect(tree).toMatchSnapshot()
        expect(tree).toHaveStyleRule('flex', '3 0 50%')
        expect(
            renderer.create(<Box flex={{ basis: '42.2%' }}>something</Box>).toJSON()
        ).toHaveStyleRule('flex', '1 1 42.2%')
    })
    it('allows hacking sizes', () => {
        SIZES.large = '103px'
        SCREEN_SIZES.lg = 'min-width: 1001px'
        const box = renderer.create(<Box margin="top" justify={{ lg: 'start' }} pad={{ top: 'large' }}>l</Box>)
        const tree = box.toJSON()
        expect(tree).toHaveStyleRule('padding-top', '103px')
        expect(tree).toHaveStyleRule('justify-content', 'flex-start', {
            media: `@media (min-width: 1001px)`,
        })
        expect(box).toMatchSnapshot();
    })

    it('renders using `as`', () => {
        const box = renderer.create(<Box as="button">btn</Box>).toJSON() as any
        expect(box['type']).toEqual('button')
    })

    it('is responsive', () => {
        const box = renderer.create(<Box justify={{ lg: 'evenly' }} align={{ md: 'stretch' }} alignContent={{ sm: 'end', lg: 'between' }}>btn</Box>)
        expect(box).toMatchSnapshot()
        const tree = box.toJSON()
        expect(tree).toHaveStyleRule('align-items', 'stretch', {
            media: `@media (${SCREEN_SIZES.md})`,
        })
        expect(tree).toHaveStyleRule('justify-content', 'space-evenly', {
            media: `@media (${SCREEN_SIZES.lg})`,
        })
        expect(tree).toHaveStyleRule('align-content', 'space-between', {
            media: `@media (${SCREEN_SIZES.lg})`,
        })
    })
})
