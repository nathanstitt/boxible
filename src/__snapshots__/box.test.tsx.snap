// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box Component allows hacking sizes 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  outline: none;
  max-width: 100%;
  padding-top: 103px;
  margin-top: .5rem;
}

<div
  className="emotion-0"
>
  l
</div>
`;

exports[`Box Component renders and matches snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  outline: none;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: flex-end;
  -ms-flex-line-pack: flex-end;
  align-content: flex-end;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  flex-basis:50% justify-content: space-evenly;
  gap: .8rem;
  min-height: 20%;
  max-width: 80%;
  height: 100%;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 1rem;
  margin: .2rem;
}

<div
  className="a-box-for-testing emotion-0"
  id="1"
>
  A test
</div>
`;
